### DISCUSSION TITLE
bass_boost

### DISCUSSION THREAD

**Discussion by Sarabanga** [savonet/liquidsoap#2165] [User Support]
bass_boost does not work as written in comments.
It is clear from the description that it should work like "_Frequency below which sound is considered as bass_", It is like "low shelf" filter,
but we have a completely different sound.

**Comment by smimram** [comment]
Thanks for the report, it seems to be due to quite a bad bug. I have minimized to this:
```
s = sine()
s' = amplify(0.001, s)
s = add(normalize=false, [s, s'])

output(s)
```
We hear almost nothing wereas we should hear one sine (or more precisely 1.001 sine).

@toots any idea? Looks like a bug with the sharing of frames...

**Comment by toots** [comment]
@Sarabanga is the bug happening on `main` or in `2.0.2`? `main` has been in a state of flux, with some heavy changes already committed and others pending which I haven't had time to complete yet.

**Comment by Sarabanga** [comment]
@toots , 2.0.2.
liquidsoap_2.0.2-ubuntu-focal-1_amd64.deb
Ubuntu 20.04

**Comment by toots** [comment]
Thank you. We'll have a look then.

**Comment by smimram** [comment]
For future reference, here is a more full-fledged test:
```
#!../../src/liquidsoap ../../libs/stdlib.liq

%include "test.liq"

s = sine()
s = add(normalize=false, [s, amplify(0., s)])
s = rms(s)
rms = s.rms
output(s)

def on_done () =
  print("RMS: #{rms()}")
  if rms() > 0.5 then test.pass() else test.fail() end
  shutdown()
end

thread.run(delay=2., on_done)
```

**Comment by smimram** [comment]
For info 541831195ac1f8eb64fa76ff3db096c1ea59c649 is passing the test. Let's bissect.

**Comment by smimram** [comment]
The outcome is, as expected: 7057092d52adcfb5b0d42f3cf24a3709d8fd8e07 is the culprit!...

**Comment by toots** [comment]
Ha! But the original reporter is also pointing to `2.0.2`. Could we have two different issues here?

**Comment by smimram** [comment]
Ah right, it could, my test is passing with 2.0.2...

**Comment by smimram** [comment]
Ok so here you go, bass boost with increasing gain in 2.0.2. The effect starts to get really noticeable starting from 0 dB, you can hear that there is mostly no distortion in the higher frequencies:

https://user-images.githubusercontent.com/2012073/150294732-609b6542-b40a-4c2a-8c5e-d331fd045b4e.mp4

So, it seems to work in 2.0.2 for me.

For info, the script I used is
```
def gain()
  -20. + time.up()
end

s = single("annotate:liq_cue_in=50.:em.mp3")
s = cue_cut(s)
s = bass_boost(gain=gain, s)
s = mux_video(video=blank(), s)
s = mksafe(s)

s = video.add_text.native({"Bass boost:\n#{string_of_float(decimal_places=0, gain())} dB"}, s, speed=0, size=100, y=260)
output.audio_video(s)
enc = %ffmpeg(format="mp4",
%audio(codec="libmp3lame", q=4),
%video(codec="libx264", preset="fast", crf=20))
output.file(enc, "/tmp/out.mp4", s)
```

**Comment by toots** [comment]
@Sarabanga how does @smimram's test compares to you?

**Comment by Sarabanga** [comment]
It sounds good.
my current sound is like a beaver singing from a toilet bowl.

**Comment by toots** [comment]
Ha 😅

Can you share a script to reproduce?

**Comment by Sarabanga** [comment]
All by default.

radio = amplify(1.,override="replay_gain",radio)
radio1 = sky(radio)
radio2 = nrj(radio)
radio3 = bass_boost(radio1)

Listen here
http://kulverstukas.nethub.fi:8000/

**Comment by Sarabanga** [comment]
beaver comes after 6db
It is not "low shelf"

https://user-images.githubusercontent.com/96509626/150540481-659b7339-cc0c-4280-b6bf-f45a1579f48a.mp4

**Comment by toots** [comment]
Thanks for checking on this, @Sarabanga. Sam and myself cannot really hear it on our systems and also with headsets but I think I have a lead:

<img width="1000" alt="Screen Shot 2022-01-21 at 9 20 27 AM" src="https://user-images.githubusercontent.com/871060/150552649-4f726060-753c-4bf5-b8e1-e3738eeff876.png">

It looks like the signal starts clipping after a certain threshold, which might be responsible for the artifact that you are hearing.

My guess is that, in order to do. automated audio processing like that, you'd need to make sure that you have enough room in your signal on each track, which seems tricky. As such, you'd probably need to at least:
* Normalize to -6dB
* Apply bass_boost to +6dB
* Renormalize after the filter.

Another approach could be to apply the bass_boost before the sky and nrj which, typically, already push your loudness to the max, leaving very little room for additional boost.

I'm not sure that this is actually an issue with us so I'm converting this to a discussion. Thanks for reporting, tho, this is interesting info!

**Comment by Sarabanga** [comment]
thanks for the advice!
